### DISCUSSION TITLE
Voodoo4/5 and SoftGPU

### DISCUSSION THREAD

**Discussion by Torinde** [bochs-emu/Bochs#306] [Ideas]
What is Bochs currently using as maximum number of threads for the Voodoo3 emulation?
If you enable more threads (readily available on modern CPUs), then probably it'll be possible to reach **Voodoo5** levels (see [here](https://github.com/PCBox/PCBox/discussions/89)). 
- Materials that may help: [voodoo45_spec.pdf](https://github.com/86Box/86Box/files/9740186/voodoo45_spec.pdf), [voodoo45_spec.doc](https://vgamuseum.info/images/doc/3dfx/voodoo45_spec.doc), [voodoo45_databook.pdf](https://vgamuseum.info/images/doc/3dfx/voodoo45_databook.pdf), Voodoo5 [ROM v5500.zip](https://github.com/86Box/86Box/files/14893805/ROM.v5500.zip) (Voodoo5 6000 has a [modern incarnation](https://www.modlabs.net/forum/topic/60674/), so surely its ROM was also dumped)
#

Another avenue for increasing graphics performance is with hypervisor drivers like [SoftGPU](https://github.com/JHRobotics/softgpu) (Win9x and potentially NT/XP) and [viogpu3d](https://github.com/virtio-win/kvm-guest-drivers-windows/pull/943) (Win8.1+) - can Bochs implement:
- VMWare SVGA-II
- QEMU [virtio vga](https://www.kraxel.org/blog/2019/09/display-devices-in-qemu/#virtio-vga) (virgl) - and maybe even expand it with venus/Vulkan

**Comment by Vort** [comment]
> What is Bochs currently using as maximum number of threads for the Voodoo3 emulation?

1 thread as far as I know.

> If you enable more threads

I suspect it will require large amount of changes and consequently create lots of new bugs.
Meanwhile Voodoo support already have lots of bugs, which wait for fixing for years.
It may be better to port fixes from MAME first, and then start to increase complexity.
